File: 1132.js
const fs = require('fs');

function solve() {
	const input = fs.readFileSync('/dev/stdin').toString().split('\n');
	let inputIdx = 0;

	const N = +input[inputIdx++];
	let alpha = {};
	const start = new Set();
	const power = [1];
	for (let i = 0; i < 11; i++) {
		power.push(power[i] * 10);
	}
	for (let i = 0; i < N; i++) {
		const tmp = input[inputIdx++];
		start.add(tmp[0]);
		for (let j = 0; j < tmp.length; j++) {
			alpha[tmp[j]] = (alpha[tmp[j]] || 0) + power[tmp.length - j - 1];
		}
	}

	alpha = Object.entries(alpha).sort((a, b) => b[1] - a[1]);
	if (alpha.length === 10 && start.has(alpha[9][0])) {
		let idx = 8;
		while (start.has(alpha[idx][0])) idx--;
		alpha.push(...alpha.splice(idx, 1));
	}
	let ans = 0;
	let num = 9;
	for (const [_, digit] of alpha) {
		ans += num * digit;
		num--;
	}
	return ans;
}

console.log(solve());

File: 1132.py
from sys import stdin

input = stdin.readline


def solve():
    N = int(input())
    power = [1]
    for _ in range(11):
        power.append(10 * power[-1])

    alpha = {}
    start = set()
    for _ in range(N):
        tmp = input()[:-1]
        start.add(tmp[0])
        for i in range(len(tmp)):
            alpha[tmp[i]] = alpha.get(tmp[i], 0) + power[len(tmp) - i - 1]

    alpha = sorted(alpha.items(), key=lambda x: -x[1])
    if len(alpha) == 10 and alpha[-1][0] in start:
        idx = -2
        while alpha[idx][0] in start:
            idx -= 1
        alpha.append(alpha.pop(idx))

    ans = 0
    num = 9
    for _, digit in alpha:
        ans += num * digit
        num -= 1

    return ans


if __name__ == '__main__':
    print(solve())